### DISCUSSION TITLE
PhysicsServer and RigidBody3D gives similar performance results

### DISCUSSION THREAD

**Discussion by PanHoracy** [godot-jolt/godot-jolt#601] [General]
So, I was trying to see how much more powerful using PhysicsServer would be compared to RigidBody3D. I have implemented simple test that I have seen many times, when people compare Godot and Unity in that regard. And to my surprise, the difference is tiny. At best, I can spawn up to couple hundred more cubes (300-400 ~ 20% in the test) and I saw people claiming that they can spawn up to 10_000 of them using Server, so I assume it's something on my side. Can anyone tell my what I'm doing wrong here? Maybe it's some bug, that triggers on my side specifially?

Anyway, there is a projet:
[PhysicsTest2.zip](https://github.com/godot-jolt/godot-jolt/files/12701100/PhysicsTest2.zip)

**Comment by mihe** [comment]
Thank you for sharing the project.

> I saw people claiming that they can spawn up to 10_000 of them using Server

Keep in mind that the scene @Bimbam360' showed [on Twitter](https://twitter.com/Bimbam_tm/status/1705140967312097492) had bodies that were all/mostly stationary, which lets the bodies go to sleep, which in turn cuts the simulation time dramatically.

In your scenes everything gets jumbled up together, which I would imagine prevents most of the bodies from going to sleep. If you disable VSync and then wait for a little bit after the bodies have stopped spawning and the scene has settled you'll note that the FPS shoots up dramatically.

But while that part doesn't surprise me much, I am still surprised by just how small the bump in performance is in your scene. I'm also seeing roughly a 15% difference on my machine, going from 3500 bodies with the nodes to 4000 with the server. The results were surprisingly inconsistent when using `Engine.get_frames_per_second()` though, but that was easily remedied by sampling/averaging `Time.get_ticks_usec()` in `_process` instead, similar to how Debug Menu does it.

Anyway, I'm curious what the frame breakdown is when using the server method, so I'll do some more in-depth profiling later.

**Comment by PanHoracy** [comment]
Yeah, I know what you are talking about. When caliming that some claim it can run 10_000 bodies I was thinking about that tweet ( https://twitter.com/H3XAGON3ST/status/1704602513360482684?t=UzpVSOJVB7_vxG3CrKxgHA&s=19) rather. And that's not even thing that I want to investigate, what number of bodies can be handled can vary, dependent on hardware and task. My only concern is that using Servers which shuld give significant performance increse is so ineffective.

**Comment by PanHoracy** [comment]
I don't know if that makes much sense to check, but I recreated the test in Godot 3.5.2. The exact same setup, and I found out that Server implementation performed WORSE there, in both bullet and GDP. In bullet, it was around 1450 to 1400, and on GDP 1050 to 950. I also checked profiler, and scripts were not the bottleneck, but physics time. I don't really know if that indicates anything, but it was interesting to uncover.
